src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -78,8 +186,41 @@ KafkaSink::KafkaSink(const Kafka * kafka, const Block & header, ContextPtr conte
         }
     }
 
-    rd_kafka_conf_set_dr_msg_cb(conf, [](rd_kafka_t * /* producer */, const rd_kafka_message_t * msg, void * /* opaque */) {
-        static_cast<WriteBufferFromKafka *>(msg->_private)->onMessageDelivery(msg);
+    auto * topic_conf = rd_kafka_conf_get_default_topic_conf(conf);
+    if (!topic_conf)
+    {
+        topic_conf = rd_kafka_topic_conf_new();
+        /// conf will take the ownership of topic_conf
+        rd_kafka_conf_set_default_topic_conf(conf, topic_conf);
+    }
+
+    rd_kafka_conf_set_opaque(conf, this);
+    rd_kafka_topic_conf_set_opaque(topic_conf, this);
+
+    /// With partitioner callback, we can get the up-to-date partition count w/o additional effort.
+    rd_kafka_topic_conf_set_partitioner_cb(topic_conf, [](const rd_kafka_topic_t * /*rkt*/,
```

Comment:
This look another dangling lambda object which get dtor-ed and still referenced in C lib. We will need hold on to the call lambda callback or refactor it out as a free function just as we have done in `KafkaLog` for all callbacks.

Cross language / lib  memory passing deserves great attentions.

Reply:
I am still getting used to c++ :)